File: examples/alamo_snapsPlot.py
#!/usr/bin/env python3

import os
import sys
import glob
import numpy as np
import matplotlib.pyplot as plt


if __name__ == "__main__":

	# Check input parameters
	if len(sys.argv) != 3:
		print("alamo_snapsPlot.py:: error: Invalid argument list")
		print(" ")
		print("Usage:")
		print(" ")
		print("  ./alamo_snapsPlot.py <snaps_path> <images_path>")
		print(" ")
		print("by: Mauri Favaron - 2018")
		sys.exit(1)
	snapsPath  = sys.argv[1]
	imagesPath = sys.argv[2]
	
	# Find snaps in snaps path
	snaps = glob.glob(os.path.join(snapsPath, "snap_*.csv"))
	if len(snaps) <= 0:
		print("alamo_snapsPlot.py:: error: Snaps path contains no snaps")
		sys.exit(2)
	try:
		guideFile = os.path.join(snapsPath, "guide.txt")
		f = open(guideFile, "r")
		lines = f.readlines()
		f.close()
	except:
		print("alamo_snapsPlot.py:: error: No or invalid guide file")
		sys.exit(3)
	if len(lines) <= 0:
		print("alamo_snapsPlot.py:: error: No or invalid guide file")
		sys.exit(3)
		
	# Get plotting extrema from the guide file
	data = np.loadtxt(guideFile)
	xmin = data[0]
	xmax = data[1]
	ymin = data[2]
	ymax = data[3]
	zmin = data[4]
	zmax = data[5]
	dx   = data[6]
	dy   = data[7]
	amax = data[8]
	
	# Process snapshot files
	for snap in sorted(snaps):
	
		# Form output image name
		snapName  = os.path.basename(snap).replace('csv', 'png')
		imageName = os.path.join(imagesPath, snapName)
		print("Processing " + snap)
		
		# Get data from snapshot, and retrieve particles coordinates from it
		snapData = np.loadtxt(snap, skiprows=1, delimiter=",")
		shp = snapData.shape
		if snapData.size > 0:
			if len(shp) > 1:
				xp       = snapData[:,0]
				yp       = snapData[:,1]
				ap       = snapData[:,7] / amax
			else:
				xp       = []
				yp       = []
				ap       = []
		else:
			xp       = []
			yp       = []
			ap       = []
		
		# Plot data
		fig = plt.figure()
		ax = fig.add_subplot(111, aspect='equal')
		levels = np.array([(0,0,0,a) for a in ap])
		ax.scatter(xp, yp, s=0.1, color=levels)
		plt.xlim(xmin,xmax)
		plt.ylim(ymin,ymax)
		plt.savefig(imageName)
		plt.close()
		

File: examples/alamo_SnapsPlot.jl
#!/usr/bin/env julia

import Pkg
#Pkg.add("Plots")
using Plots

struct Particle

	x::Float32
	y::Float32
	z::Float32
	q::Float32
	t::Float32
	
	# Constructor (uninitialized)
	Particle() = new()
	
end


# Get parameters
if length(ARGS) != 2

	println("alamo_SnapPlot.jl - Program generating particles scatter plots for movie production")
	println()
	println("Usage:")
	println()
	println("    ./alamo_SnapPlot.jl <SnapDataPath> <SnapPlotPath>")
	println()
	println("This file is part of the pbl_met project.")
	exit(1)
	
end
sSnapPath = ARGS[1]
sOutPath  = ARGS[2]

# Get boundary information
line = open(sSnapPath * "/guide.txt") do f
	lines = readlines(f)
	(lines[1])
end
blocks = split(line)
xmin   = parse(Float64, blocks[1])
xmax   = parse(Float64, blocks[2])
ymin   = parse(Float64, blocks[3])
ymax   = parse(Float64, blocks[4])
zmin   = parse(Float64, blocks[5])
zmax   = parse(Float64, blocks[6])
dx     = parse(Float64, blocks[7])
dy     = parse(Float64, blocks[8])
tmax   = parse(Float64, blocks[9])
dt     = parse(Float32, blocks[10])
year   = parse(Int32, blocks[11])
month  = parse(Int32, blocks[12])
day    = parse(Int32, blocks[13])
hour   = parse(Int32, blocks[14])
minute = parse(Int32, blocks[15])
second = parse(Float64, blocks[16])

# Iterate over input path contents, and process them
files = readdir(sSnapPath)
for file in files

	# Get file as it is
	sFileName = sSnapPath * "/" * file
	nData = filesize(sFileName) / (5*4)
	data = read(sFileName, Particle, nData)
	
	# Build scatter plot of all points within boundary, which result in "active" state
	
	# Inform users about advancement
	println("Just processed:" * sFileName)
	
end
